lib/metasploit/framework/data_service/stubs/host_data_service.rb

```diff
@@ -27,4 +27,16 @@ def update_host(opts)
   def delete_host(opts)
     raise 'HostDataService#delete_host is not implemented'
   end
-end
\ No newline at end of file
+
+  def get_host_tags(opts)
```

Comment:
I realise you're just following the convention here but what's the purpose of this class?

Reply:
From what I can tell it's to ensure these methods actually get implemented by any Data Services in use. Idea is if a dev adds a new Data Service and forgets to implement one of the methods in the stub classes, a human readable error gets thrown.

Reply:
The method you commented on is simply a stub for the actual `get_host_tags` implementation. As for the `HostDataService` module, it is one of the stub modules that are used to define the interface to be implemented, see `lib/metasploit/framework/data_service.rb`.
> All data service implementations should include this module to ensure proper implementation

Reply:
ok then...I think I understand that, I guess what I don't get is how things break if you don't implement one of the stubs, maybe my search skills are lacking but I can't see where it gets used

Reply:
It gets `included` in the DataService module, which also contains the DataProxy class. `framework.db` is set as a DataProxy instance, and the DataProxy forwards any method calls onto the DataService in use. So when we call `framework.db.example_method`, the DataProxy passes the call to `example_method` onto the "current" DataService. If the DataService hasn't implemented `example_method`, it's implementation in one of the stubs get's called instead.